File: Core/CLR/System/Aspect/TypeAspectAttribute.cs
using System;
using Bridge.CLR;

namespace Bridge.Aspects
{
    /// <summary>
    /// 
    /// </summary>
    [Ignore]
    [Name("Bridge.Aspects.TypeAspectAttribute")]
    [AttributeUsage(AttributeTargets.Assembly | AttributeTargets.Class, AllowMultiple=true)]
    public abstract class TypeAspectAttribute : MulticastAspectAttribute
    {
        public const string Format = "new {0}({3}).init({2});";
        public const string MergeFormat = "Bridge.merge(new {0}({3}), {{{4}}}).init({2});";

        public readonly string TypeName;
        public readonly object Instance;

        public virtual void OnInstance(TypeAspectEventArgs eventArgs)
        {
        }

        public virtual void OnAfterInstance(TypeAspectEventArgs eventArgs)
        {
        }

        protected virtual bool RunTimeValidate(object instance)
        {
            return true;
        }
    }

    [Ignore]
    [Name("Object")]
    public class TypeAspectEventArgs
    {
        public readonly object Instance;
        public readonly string TypeName;
        public readonly object[] Arguments;
    }
}

File: Core/CLR/resources/aspect/TypeAspectAttribute.js
Bridge.Class.extend('Bridge.Aspects.TypeAspectAttribute', {
    $extend: [Bridge.Aspects.MulticastAspectAttribute],

    onInstance: Bridge.emptyFn,
    onAfterInstance: Bridge.emptyFn,

    init: function (instance, arguments) {
        this.instance = instance;
        this.typeName = instance.$$name;
        this.arguments = arguments;

        if (!this.runTimeValidate(instance)) {
            return;
        }

        this.instance.$$aspects = this.instance.$$aspects || {};
        if (!this.instance.$$aspects[this.$$name]) {
            this.instance.$$aspects[this.$$name] = [];
        }
        this.instance.$$aspects[this.$$name].push(this);
        this.$$setAspect();
    },

    $$setAspect: function () {
        var me = this;
        this.$$targetInitCtor = this.instance.$$initCtor;
        this.instance.$$initCtor = function () {
            var args = {
                instance: me.instance,
                typeName: me.typeName,
                arguments: arguments
            };

            me.onInstance(args);

            me.$$targetInitCtor.apply(me.instance, args.arguments)

            me.onAfterInstance(args);
        };
    },

    runTimeValidate: function () {
        return true;
    }
});